File: zscala/src/main/scala/zingg/zs/Phase.scala
package zingg.zs

/** The Zingg pipeline phases recognised by the CLI. Same names as the legacy
  * `zingg.common.client.options.ZinggOptions` so existing `zingg.sh` scripts
  * keep working.
  */
sealed abstract class Phase(val name: String)

object Phase {
  case object Train            extends Phase("train")
  case object Match            extends Phase("match")
  case object TrainMatch       extends Phase("trainMatch")
  case object FindTrainingData extends Phase("findTrainingData")
  case object Label            extends Phase("label")
  case object Link             extends Phase("link")
  case object GenerateDocs     extends Phase("generateDocs")
  case object Recommend        extends Phase("recommend")
  case object UpdateLabel      extends Phase("updateLabel")
  case object FindAndLabel     extends Phase("findAndLabel")

  val all: Seq[Phase] = Seq(
    Train, Match, TrainMatch, FindTrainingData, Label,
    Link, GenerateDocs, Recommend, UpdateLabel, FindAndLabel
  )

  def parse(s: String): Either[String, Phase] =
    all.find(_.name == s).toRight(
      s"unknown phase '$s'. Valid: ${all.map(_.name).mkString("|")}"
    )
}

File: python/zpy/src/zingg_py/phase.py
"""Pipeline phases — port of ``Phase.scala``."""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class Phase:
    name: str


Phase.Train = Phase("train")
Phase.Match = Phase("match")
Phase.TrainMatch = Phase("trainMatch")
Phase.FindTrainingData = Phase("findTrainingData")
Phase.Label = Phase("label")
Phase.Link = Phase("link")
Phase.GenerateDocs = Phase("generateDocs")
Phase.Recommend = Phase("recommend")
Phase.UpdateLabel = Phase("updateLabel")
Phase.FindAndLabel = Phase("findAndLabel")

ALL = [
    Phase.Train, Phase.Match, Phase.TrainMatch, Phase.FindTrainingData, Phase.Label,
    Phase.Link, Phase.GenerateDocs, Phase.Recommend, Phase.UpdateLabel, Phase.FindAndLabel,
]


def parse(s: str) -> Phase:
    for p in ALL:
        if p.name == s:
            return p
    valid = "|".join(p.name for p in ALL)
    raise ValueError(f"unknown phase '{s}'. Valid: {valid}")
